presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java

```diff
@@ -0,0 +1,88 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.prestosql.plugin.kafka.encoder.EncoderFieldValueProvider;
+import io.prestosql.plugin.kafka.encoder.RowEncoder;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.connector.ConnectorPageSink;
+import io.prestosql.spi.type.Type;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class KafkaPageSink
+        implements ConnectorPageSink
+{
+    private final String topicName;
+    private final List<KafkaColumnHandle> columns;
+    private final List<Type> columnTypes;
+    private final KafkaProducerFactory producerFactory;
+    private final RowEncoder keyEncoder;
+    private final RowEncoder messageEncoder;
+
+    public KafkaPageSink(
+            KafkaProducerFactory producerFactory,
+            String topicName,
+            List<KafkaColumnHandle> columns,
+            RowEncoder keyEncoder,
+            RowEncoder messageEncoder)
+    {
+        this.producerFactory = requireNonNull(producerFactory, "producerFactory is null");
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.columns = ImmutableList.copyOf(requireNonNull(columns, "columnNames is null"));
+        this.keyEncoder = requireNonNull(keyEncoder, "keyEncoder is null");
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+
+        this.columnTypes = columns.stream()
+                .map(KafkaColumnHandle.class::cast)
+                .map(KafkaColumnHandle::getType)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
+    {
+        try (KafkaProducer<byte[], byte[]> producer = producerFactory.create()) {
+            List<EncoderFieldValueProvider> row = new ArrayList<>(columnTypes.size());
+            for (int position = 0; position < page.getPositionCount(); ++position) {
+                for (int channel = 0; channel < page.getChannelCount(); ++channel) {
+                    row.add(new KafkaFieldValueProvider(columns.get(channel), page.getBlock(channel), position));
```

Comment:
We should segregate key columns (if any) and use the `keyEncoder` to generate the key.  You can probably tackle this and dealing with internal columns as a follow-up PR (no need to make this one larger than it already is).